### DISCUSSION TITLE
New user needs startup documentation

### DISCUSSION THREAD

**Discussion by RedKing7-GA** [libMesh/libmesh#3913] [General]
I'm a new user interested in using libmesh to read meshes and manipulate them.  Where can I find documentation that will walk me through using the various "functions" to do this?

**Comment by roystgnr** [comment]
Sorry for letting this notification go unnoticed.

There's no simple walkthrough, but [the first introduction example](https://github.com/libMesh/libmesh/blob/master/examples/introduction/introduction_ex1/introduction_ex1.C) creates a `Mesh` object, reads its data from file, and writes to another file.  The [meshtool app](https://github.com/libMesh/libmesh/blob/master/src/apps/meshtool.C) has a few modification options in between.  The [MeshModification namespace](https://github.com/libMesh/libmesh/blob/master/include/mesh/mesh_modification.h) has a few other modification options encapsulated.  For any more custom manipulation you'll want to iterate over nodes and/or elements yourself; see the other introduction example codes for a few iterator ranges.  The most up-to-date build of the libMesh Doxygen documentation is probably [the one hosted at the MOOSE Framework site.](https://mooseframework.inl.gov/docs/doxygen/libmesh/classes.html)

**Comment by RedKing7-GA** [comment]
I've taken a brief look at this, but still can't figure out what I need.  I just want to read an existing mesh, carve it up into a number of smaller subdomains and read the subdomain descriptions.  I didn't see where this could be done with DistributedMesh; I see nothing in the header file that gives me a clue as to how to do this.

**Reply by roystgnr** [reply]
`DistributedMesh` vs `ReplicatedMesh` isn't about how the mesh is carved up on a conceptual level (e.g. here's the subdomain where the solid mechanics get solved, there's the one where the fluid mechanics get solved), but rather on a software level (e.g. here's the partition subset where MPI rank 0 does calculations, there's the subset where MPI rank 1 does).  "Distributed" just means that neither processor can see all the mesh (saving memory and usually CPU time at the cost of more complicated code and inter-processor communication).

I'm not sure what you mean by "subdomain descriptions".  We support "subdomain names" internally and with a number of file formats, and I guess those can be arbitrary-length strings, but in practice everybody uses names like "fluid" and "solid", not "In this essay I will demonstrate that..."  But if those are what you need, each `Elem` has a `subdomain_id()` that lets you read and set which integer id associated with a subdomain applies to that element, and a `Mesh` (look in the `MeshBase` APIs) lets you read and set a `subdomain_name()` for each id.

**Comment by RedKing7-GA** [comment]
Let me clarify. 

First, I don't intend to do any type of "mechanics" with the mesh. 

 I simply want to take an unstructured mesh with lots of elements and divided it into smaller, contiguous meshes.  I want to retrieve the node locations, connectivity information, and nearest neighbor information for each of the smaller meshes.

**Reply by roystgnr** [reply]
Node locations are `MeshBase::point()`; connectivity is `Elem::neighbor_ptr()` or `Elem::node_ref()` depending on which kind you mean.  To create a new smaller mesh from a subset of a larger one your best bet is to iterate over the larger one (e.g. `MeshBase::element_ptr_range()`) and copy (`MeshBase::add_point()`, `MeshBase::add_elem()`) into the appropriate smaller one as desired, then `MeshBase::prepare_for_use()` on each smaller one to reconstruct things like neighbor pointers that can be inferred without copying.
